catalog/core/catalog-core-commands/src/main/java/org/codice/ddf/commands/util/DigitalSignature.java

```diff
@@ -0,0 +1,130 @@
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PrivilegedAction;
+import java.security.Signature;
+import java.security.SignatureException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.Certificate;
+import org.apache.commons.io.IOUtils;
+import org.codice.ddf.security.common.Security;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class DigitalSignature {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(DigitalSignature.class);
+
+  private static final int BUFFER_SIZE = 1024;
+
+  private static final int OFFSET = 0;
+
+  private KeyStore keyStore;
+
+  public DigitalSignature() {
+    Security security = Security.getInstance();
+    this.keyStore =
+        AccessController.doPrivileged((PrivilegedAction<KeyStore>) security::getSystemKeyStore);
+  }
+
+  public DigitalSignature(KeyStore keyStore) {
+    this.keyStore = keyStore;
+  }
+
+  public byte[] createDigitalSignature(InputStream data, String alias, String password)
+      throws IOException {
+    PrivateKey privateKey = getPrivateKey(alias, password);
+
+    if (privateKey == null) {
+      throw new CatalogCommandRuntimeException("Unable to retrieve private key");
+    }
+
+    try {
+      Signature rsa = Signature.getInstance("SHA256withRSA");
+
+      rsa.initSign(privateKey);
+
+      byte[] buffer = new byte[BUFFER_SIZE];
+      int len;
+
+      while ((len = data.read(buffer)) >= 0) {
+        rsa.update(buffer, OFFSET, len);
+      }
+
+      return rsa.sign();
+    } catch (SignatureException | InvalidKeyException | NoSuchAlgorithmException e) {
+      String message = "An error occurred while signing file";
+      LOGGER.debug(message, e);
+      throw new CatalogCommandRuntimeException(message, e);
+    }
+  }
+
+  public boolean verifyDigitalSignature(
+      InputStream data, InputStream signature, String certificateAlias) throws IOException {
+    byte[] sigToVerify = IOUtils.toByteArray(signature);
+
+    Certificate certificate = getCertificate(certificateAlias);
+
+    if (certificate == null) {
+      throw new CatalogCommandRuntimeException("Unable to retrieve certificate");
+    }
+
+    try {
+      Signature rsa = Signature.getInstance("SHA256withRSA");
+      rsa.initVerify(certificate);
+
+      byte[] buffer = new byte[BUFFER_SIZE];
+      int len;
+
+      while ((len = data.read(buffer)) >= 0) {
+        rsa.update(buffer, OFFSET, len);
+      }
+
+      return rsa.verify(sigToVerify);
+    } catch (InvalidKeyException | SignatureException | NoSuchAlgorithmException e) {
+      String message = "An error occurred while verifying file";
+      LOGGER.debug(message, e);
+      throw new CatalogCommandRuntimeException(message, e);
+    }
+  }
+
+  private PrivateKey getPrivateKey(String alias, String password) {
+    try {
+      return (PrivateKey) keyStore.getKey(alias, password.toCharArray());
```

Comment:
❓ Do we know for sure this will always return a `PrivateKey`?

Reply:
We cannot be certain that it will be a private key. Changed